monad-chronicle/src/Control/Monad/Trans/Chronicle/CPS.hs

```diff
@@ -0,0 +1,272 @@
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE Trustworthy           #-}
+{-# LANGUAGE UndecidableInstances  #-}
+-----------------------------------------------------------------------------
+-- | Module     :  Control.Monad.Chronicle
+--
+-- Hybrid error/writer monad class that allows both accumulating outputs and
+-- aborting computation with a final output.
+--
+-- The expected use case is for computations with a notion of fatal vs.
+-- non-fatal errors.
+
+-----------------------------------------------------------------------------
+module Control.Monad.Trans.Chronicle.CPS (
+    -- * The Chronicle monad
+    Chronicle, chronicle, runChronicle,
+    -- * The ChronicleT monad transformer
+    ChronicleT(..),
+    -- * Chronicle operations
+    dictate, disclose, confess,
+    memento, absolve, condemn,
+    retcon,
+    ) where
+
+import Control.Applicative
+import Control.Monad
+import Control.Monad.Fix
+import Control.Monad.Trans
+import Data.Default.Class
+import Data.Functor.Identity
+import Data.Semigroup
+
+import Control.Monad.Error.Class
+import Control.Monad.Reader.Class
+import Control.Monad.RWS.Class
+import Data.These
+import Data.These.Combinators     (mapHere)
+import Prelude
+
+#ifdef MIN_VERSION_semigroupoids
+import Data.Functor.Apply (Apply (..))
+import Data.Functor.Bind  (Bind (..))
+#endif
+
+-- --------------------------------------------------------------------------
+-- | A chronicle monad parameterized by the output type @c@.
+--
+--   The 'return' function produces a computation with no output, and '>>='
+--   combines multiple outputs with '<>'.
+type Chronicle c = ChronicleT c Identity
+
+chronicle :: Semigroup c => Monad m => These c a -> ChronicleT c m a
+chronicle (This c')    = ChronicleT $ \c -> let ct = c <> c' in ct `seq` pure (This ct)
+chronicle (That     a) = ChronicleT $ \c -> pure (These c a)
+chronicle (These c' a) = ChronicleT $ \c -> let ct = c <> c' in ct `seq` pure (These ct a)
+{-# INLINE chronicle #-}
+
+runChronicle :: Monoid c => Chronicle c a -> These c a
+runChronicle = runIdentity . flip runChronicleT mempty
```

Comment:
I was afraid there is something like that, i.e. `Monoid` pops in, instead of `Semigroup`. 

Just guessing, would `ChronicleT { runChronicleT :: Maybe c -> m (These c a) }` work better (and still do the job?)